File: 022.py
#   022
#   Generate Parentheses
#   2015-04-29
#####################################################
class Solution:
    # @param {integer} n
    # @return {string[]}
    def generateParenthesis(self, n):
        v = []
        if n == 0:
            return v
        v.append("(")
        while len(v[0]) < 2*n:
            s = v.pop(0)
            l = r = 0
            for c in s:
                if c == '(':
                    l+=1
                else:
                    r+=1
            if l<n:
                v.append(s+"(")
            if l>r:
                v.append(s+")")
        return v

#####################################################

sol = Solution()
for n in range(0,9):
    v = sol.generateParenthesis(n)
    #print v
    print n, len(v)

File: ruby/022.rb
# 022
# Generate Parentheses
# 2015-04-29
#####################################################
# @param {Integer} n
# @return {String[]}
def generate_parenthesis(n)
  v = []
  return v if n == 0
  v.push("(")
  while v.first.size < 2*n
    s = v.shift
    l = r = 0
    s.each_char do |c|
      if c == "("
        l+=1
      else
        r+=1
      end
    end

    if l<n
      v.push(s+"(")
    end
    if l>r
      v.push(s+")")
    end
  end
  v
end

#####################################################

n = 4
v = generate_parenthesis(n)
puts v
puts v.size


